File: docs/programming/programming_with_gene_annotations3/code/gff_to_sqlite.py
import os
import sqlite3
import gmsgff

def parse_arguments():
    import argparse
    parser = argparse.ArgumentParser(
        description = """Convert a GFF3 file to sqlite3 format.
        The result will be a table with the GFF3 fields, and with ID and Parent fields in columns.
        """
    )
    parser.add_argument(
        '--input',
        help ='The path of a GFF3-formatted file to work with',
        required = True
    )
    parser.add_argument(
        '--output',
        help = "The path to the output file",
        required = True
    )
    # add other needed arguments here
    return parser.parse_args()

def get_dataset_name( filename ):
    # Get the input name without directories
    result = os.path.basename( args.input )
    # Get rid of the .gff suffix - python has a function for this
    result = os.path.splitext( result )[0]
    return( result )

def process( args ):
    print( "++ process(): loading data from '%s'..." % args.input )
    data = gmsgff.parse_gff3_to_dataframe( args.input )
    print( "++ ok, loaded %d rows and %d columns of data." % ( data.shape[0], data.shape[1] ) )
    print( "first few rows are:" )
    print( data.head() )

    print( "++ process(): Adding dataset filename as a column..." )
    data.insert( loc = 0, column = 'dataset', value = get_dataset_name( args.input ) )

    print( "++ process(): Writing data to '%s', gff table..." % args.output )
    db = sqlite3.connect( args.output )
    data.to_sql( "gff", db, index = False, if_exists = "append" )

    print( "++ process(): ok, adding indexes..." )
    db.execute( "CREATE INDEX IF NOT EXISTS gff_id_index ON gff( ID )" )
    db.execute( "CREATE INDEX IF NOT EXISTS gff_parent_index ON gff( Parent )" )

    print( "++ process(): closing the database." )
    db.close()

print( "++ Welcome to gff_to_sqlite.py!" )
args = parse_arguments()

print( "++ gff_to_sqlite.py: processing..." )
process( args )

print( "++ Success!" )
print( "++ Thank you for using gff_to_sqlite.py." )

File: docs/programming/programming_with_gene_annotations3/code/gff_to_sqlite.R
library(
	tidyverse,
	quietly = TRUE
)
library( RSQLite )
library( gmsgff )

echo = function( message, ... ) {
	cat( sprintf( message, ... ))
}

parse_arguments = function() {
    library( argparse )
    parser = ArgumentParser(
        description = "Convert a GFF3 file to sqlite3 format. The result will be a table with the GFF3 fields, and with ID and Parent fields in columns."
    )
    parser$add_argument(
        '--input',
        type = "character",
        help = "The path of a GFF3-formatted file to work with",
        required = TRUE
    )
    parser$add_argument(
        '--output',
        type = "character",
        help = "The path of the output file.",
        required = TRUE
    )
    return( parser$parse_args() )
}

get_dataset_name = function( filename ) {
    # Get the input name without directories
    result = basename( filename )
    # Get rid of the .gff suffix using a 'regular expression'
    # [.] matches a dot
    # [^.]* matches any number of things that are NOT dots
    # $ matches the end of the string
    result = gsub( "[.][^.]*$", "", result )
    return( result )
}


process = function( args ) {
	echo( "++ process(): loading data from '%s'...\n", args$input )
    data = gmsgff::parse_gff3_to_dataframe( args$input )
	echo( "++ ok, loaded %d rows and %d columns of data.\n", nrow(data), ncol(data) )
	echo( "first few rows are:\n" )
	print( head( data ))

	echo( "++ process(): Adding dataset filename as a column...\n" )
    data = add_column(
        data,
        dataset = get_dataset_name( args$input ),
        .before = 1
    )

	echo( "++ process(): Writing data to '%s', gff table...\n", args$output )
    db = DBI::dbConnect(RSQLite::SQLite(), args$output )
    dbWriteTable(
        db,
        "gff",
        data,
        row.names = FALSE,
		append = TRUE
    )

	echo( "++ process(): ok, adding indexes...\n" )
    # Create the indexes
    dbExecute( db, "CREATE INDEX IF NOT EXISTS gff_id_index ON gff( ID )" )
    dbExecute( db, "CREATE INDEX IF NOT EXISTS gff_parent_index ON gff( Parent )" )

    # close the connection
    dbDisconnect( db )
	echo( "++ process(): success.\n" )
}

echo( "++ Welcome to gff_to_sqlite.R!\n" )
args = parse_arguments()

echo( "++ gff_to_sqlite.R: processing...\n" )
process( args )

echo( "++ Success!\n" )
echo( "++ Thank you for using gff_to_sqlite.R.\n" )

